windows/utils/win_install_online_updates.yml

```diff
@@ -39,17 +51,71 @@
   when:
+    - win_updates_install_result.failed is defined
+    - not win_updates_install_result.failed
     - win_updates_install_result.found_update_count is defined
-    - win_updates_install_result.found_update_count | int != 0
     - win_updates_install_result.installed_update_count is defined
-    - win_updates_install_result.found_update_count | int == win_updates_install_result.installed_update_count | int
+  block:
+    # No updates found, so no updates installed
+    - name: "Set fact of Windows online updates install succeeds"
+      ansible.builtin.set_fact:
+        win_online_updates_succeed: true
+      when: win_updates_install_result.found_update_count | int == 0
+
+    # Updates found, updates installed
+    - name: "Installed update count is equal to found update count"
+      ansible.builtin.set_fact:
+        win_online_updates_succeed: true
+        win_online_updates_installed: true
+      when:
+        - win_updates_install_result.found_update_count | int != 0
+        - win_updates_install_result.installed_update_count | int == win_updates_install_result.found_update_count | int
+
+    # Updates found, while some not installed
+    - name: "Installed update count is not equal to found update count"
+      when:
+        - win_updates_install_result.found_update_count | int != 0
+        - win_updates_install_result.found_update_count | int > win_updates_install_result.installed_update_count | int
+      block:
+        - name: "Set fact of Windows online updates install status"
+          ansible.builtin.set_fact:
+            win_online_updates_installed: true
+          when: win_updates_install_result.installed_update_count | int != 0
+
+        - name: "Set fact of not installed updates list"
+          ansible.builtin.set_fact:
+            win_updates_not_installed: "{{ win_updates_install_result.updates.values() | selectattr('installed', 'equalto', false) }}"
+        - name: "Display the not installed updates list"
+          ansible.builtin.debug: var=win_updates_not_installed
+
+        - name: "Set fact of failure messages of not installed updates"
+          ansible.builtin.set_fact:
+            win_updates_failure_msgs: "{{ win_updates_not_installed | map(attribute='failure_msg') }}"
+
+        - name: "No failure message for update"
+          ansible.builtin.debug:
+            msg: "No 'failure_msg' info for not installed updates, so will not ignore these not installed updates."
+          when: win_updates_failure_msgs | length != win_updates_not_installed | length
+
+        - name: "Get failure messages for all not installed updates"
+          when: win_updates_failure_msgs | length == win_updates_not_installed | length
+          block:
+            - name: "Set fact of failure messages can not be ignored"
+              ansible.builtin.set_fact:
+                win_updates_failure_msgs: "{{ win_updates_failure_msgs | reject('search', item) }}"
```

Comment:
If the win_updates_failure_msgs is changed, the value of `win_updates_failure_msgs | length` at line 106 will be changed, and then it will quit from this block, and never update `win_online_updates_succeed`.

You can add 2 vars to save the value of `win_updates_failure_msgs | length` and `win_updates_not_installed | length`, which can avoid this situation.

Reply:
Yeah, right, thanks for catching this.